UserRoleMapper non-null groups and metadata

albertzaharovits opened the pull request:
`UserRoleMapper.UserData` is constructed by each realm and it is used to "match" role mapping expressions that eventually supply the role names of the principal.

`null` values as group names or metadata values were not properly supported. Most of the time they were filtered out, but under some circumstances they might cause NPEs (depending of how the OIDC and SAML's underlying libraries parse purposefully crafted assertions - these libraries don't have internal `null` checks the same way that the unboundid library does).

This PR enforces non-null collection values (lists and maps) earliest in the process and utilizes java 9 unmodifiable collections API, with the goal of eliminating redundant `null` filtering and repeated wrapping in`Colllections.unmodifiable*` .

albertzaharovits committed:
WIP all but SAML

albertzaharovits committed:
Groups don't dig nulls

albertzaharovits committed:
Kerberos and LDAP metadata

albertzaharovits committed:
Oidc metadata

albertzaharovits committed:
Done

albertzaharovits committed:
Nits

elasticmachine commented:
Pinging @elastic/es-security

albertzaharovits committed:
Fix List LDAP metadata

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/saml/SamlRealm.java:
```diff
@@ -439,20 +439,20 @@ private void buildUser(SamlAttributes attributes, ActionListener<AuthenticationR
+            userMetaBuilder.put(USER_METADATA_NAMEID_FORMAT, attributes.name().format);
         }
-
+        final Map<String, Object> userMeta = Map.copyOf(userMetaBuilder);
```
Okay, looking at how `userMetaBuilder` is populated I thought that the stream API (that builds the unmodifiable map) is unsuitable.
I'll give it a try though...

albertzaharovits replied:
I originally did not understood the suggestion, sorry Yogesh!
Yes, I prefer to create the `AbstractImmutableCollection` as soon as possible to avoid the unnecessary wrappings in `Collections.unmodifiableMap` and duplicated `null` filterings later on. Plus, the goal is to pass the `userMeta` to the `UserRoleMapper.UserData` not the `User`.

tvernum commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/kerberos/KerberosRealm.java:
```diff
@@ -215,18 +215,17 @@ private void resolveUser(final String userPrincipalName, final String outToken,
 
     private void buildUser(final String username, final Map<String, Object> metadata, final ActionListener<AuthenticationResult> listener) {
-        final UserRoleMapper.UserData userData = new UserRoleMapper.UserData(username, null, Collections.emptySet(), metadata, this.config);
+        final UserRoleMapper.UserData userData = new UserRoleMapper.UserData(username, null, Set.of(), metadata, this.config);
```
Is there a reason to change this? I don't feel strongly, but I do prefer `emptySet` as being more readable in a standalone context.

albertzaharovits replied:
Yes, there is a reason. I try to be consistent and return an `AbstractImmutableCollection` for every return path in the function.

Background:

The goal of the PR is to use the new API for immutable collections. They have two advantages: improved API for building small  "unmodifiable" collections (`ImmutableCollections`) and preventing wrapping of an `unmodifiable` collection by another `unmodifiable` collection (see for example `java.util.Set#copyOf`). The "drawback" is that `null` elements are not permitted.

What I want to achieve in this PR is to construct `ImmutableCollections` instances as soon as possible (using `Set.of`, `Set.copyOf` and `Collectors.toUnmodifiableSet()` - equivalent for lists) - at the interface with another library - and then pass those instances up the call chain without further "unmodifiable" wrappings or checks for `null` elements during iterations. At the lowest level, the interface with another library, I looked at their code to check if they have `null`-checks in place; if they did I used the `Set.of` or `Set.copyOf` functions, because if they catch a `null` element and throw then we have catched a bug in the library; otherwise, if there are no `null`-checks in the library, I simply filtered out the `null` elements (they were not relevant) using a `Stream` and then used the `Collectors.toUnmodifiableSet` collector.

Unfortunately, the type system is not helping us in guaranteeing that the parameter is such an "unmodifiable" collection, so `Set.copyOf` or `List.copyOf` are to be expected in the code, between "abstraction layers", but they will not create new instances if indeed the Collection was `AbstractImmutableCollection` (so, if it was created with `Set.of` instead of `Collections.emptySet()`), so they are cheap.

tvernum commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ldap/ActiveDirectoryGroupsResolver.java:
```diff
@@ -53,16 +52,16 @@ public void resolve(LDAPInterface connection, String userDn, TimeValue timeout,
                 ignoreReferralErrors, ActionListener.wrap((filter) -> {
                     if (filter == null) {
-                        listener.onResponse(Collections.emptyList());
+                        listener.onResponse(List.of());
```
Per my comment above, I think we need to decide whether we want to switch these across the board. I don't see any reason to do so, but I'm not specifically opposed to it.

albertzaharovits replied:
Per my comment above, the only reason is to get an `AbstractImmutableCollection` instead of an `AbstractCollection`, the former avoiding a new instance creation in the `Set.copyOf` or `List.copyOf` calls.

tvernum commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/oidc/OpenIdConnectRealm.java:
```diff
@@ -414,17 +410,21 @@ static ClaimParser forSetting(Logger logger, OpenIdConnectRealmSettings.ClaimSet
                             } else {
                                 throw new SettingsException("Setting [" + RealmSettings.getFullSettingKey(realmConfig, setting.getClaim())
-                                    + " expects a claim with String or a String Array value but found a "
+                                    + " expects a claim with String or a String List value but found a "
```
I'll defer to @jkakavas on this, but I think the original message was correct. This is refering to claims in the OIDC JSON, and the JSON term for a multi-valued element is an Array, not a List.

albertzaharovits replied:
Okay, I'll revert this.

tvernum commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/oidc/OpenIdConnectRealm.java:
```diff
@@ -438,23 +438,25 @@ static ClaimParser forSetting(Logger logger, OpenIdConnectRealmSettings.ClaimSet
                             } else if (claimValueObject instanceof List == false) {
                                 throw new SettingsException("Setting [" + RealmSettings.getFullSettingKey(realmConfig, setting.getClaim())
-                                    + " expects a claim with String or a String Array value but found a "
+                                    + " expects a claim with String or a String List value but found a "
```
As above (though I wonder why we have duplicated code here)

tvernum commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/pki/PkiRealm.java:
```diff
@@ -149,9 +149,8 @@ public void authenticate(AuthenticationToken authToken, ActionListener<Authentic
-        final Map<String, Object> metadata = Collections.singletonMap("pki_dn", token.dn());
-        final UserRoleMapper.UserData userData = new UserRoleMapper.UserData(token.principal(),
-                token.dn(), Collections.emptySet(), metadata, this.config);
+        final Map<String, Object> metadata = Map.of("pki_dn", token.dn());
```
Similar to my comments on empty collections, I think we need to reach a team-wide decision on `singletonXXX` vs `of`

albertzaharovits replied:
I think it would be inconsistent to use `Collections.singletonMap` for a single entry but `Map.of` for several. Similarly, it would be inconsistent to use `Collections.emptyMap()` . So if we should use `Map.of` for two entries, then we should use it for one or the empty map, right?

tvernum commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/saml/SamlRealm.java:
```diff
@@ -439,20 +439,20 @@ private void buildUser(SamlAttributes attributes, ActionListener<AuthenticationR
+            userMetaBuilder.put(USER_METADATA_NAMEID_FORMAT, attributes.name().format);
         }
-
+        final Map<String, Object> userMeta = Map.copyOf(userMetaBuilder);
```
`Map.copyOf` requires that the source map does not contain any `null` values, and we don't know that here.
I'm pretty sure it is possible for `USER_METADATA_NAMEID_FORMAT` to have null value.

albertzaharovits replied:
> I'm pretty sure it is possible for USER_METADATA_NAMEID_FORMAT to have null value.

You're right, I will do the check.

tvernum commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/support/UserRoleMapper.java:
```diff
@@ -63,12 +60,8 @@ public UserData(String username, @Nullable String dn, Collection<String> groups,
-            this.metadata = metadata == null || metadata.isEmpty()
-                    ? Collections.emptyMap() : Collections.unmodifiableMap(new HashMap<>(metadata));
+            this.groups = Set.copyOf(groups);
+            this.metadata = metadata;
```
I don't think it is appropriate for `UserData` to just assume that every caller is passing in an immutable collection.
I believe we should either be copying or wrapping (and I don't think you can safely copy because we might have null values).

albertzaharovits replied:
> I don't think it is appropriate for UserData to just assume that every caller is passing in an immutable collection.

Since the type system is not guaranteeing it, yes it is not safe. In the worst case we can copy on every function.
I will copy it...

> I believe we should either be copying or wrapping (and I don't think you can safely copy because we might have null values).

I believe this is what this PR aims, to assure there are no `null` values?

albertzaharovits committed:
Merge branch 'master' into user_tole_mapper_take2

albertzaharovits committed:
Review

albertzaharovits commented:
@tvernum @bizybot I have addressed your review, but I believe some general notes on the implementation are required.

* We should remove `null`s from collections as soon as possible, during the first construction. If we tolerate them at all, then we simply increase the surface for where the `null` filtering is required (when the collection is finally iterated upon).
* If we were to use `Map.of`, `Set.of` and `List.of` for "small" Collections, then I believe we should use the same methods for singletons and empty collections. Besides consistency, there is a small optimization in the new library that avoids repeated wrappings during the `copyOf` calls (it returns the same instance if it is unmodifiable) that are characteristic of `Collections.unmodifiable*`. So if we use the old APIs the first `copyOf` will create a new instance.
* We should not be afraid of using `Set.copyOf` because it might throw NPEs if the collection contains `null`s, to the same extent that we are not afraid of calling a method on a parameter without first `null` checking it. We should work on guaranteeing we don't use `null` elements in collections, rather than sprinkle `null` filterings to be safe, the same way we don't allow for `null` parameters unless they are annotated (or at least we presume parameters are non-`null`).

Let me know what you think!

albertzaharovits commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into user_tole_mapper_take2

albertzaharovits committed:
OOOPS native users store , uses null metadata :(

albertzaharovits committed:
Saml bug

albertzaharovits committed:
Avoid redundant copyOf

albertzaharovits committed:
Merge branch 'master' into user_tole_mapper_take2

albertzaharovits commented:
@elasticmachine run elasticsearch-ci/2

albertzaharovits committed:
Mrrrrr

albertzaharovits commented:
@elasticmachine test this please

albertzaharovits commented:
@elasticmachine run  elasticsearch-ci/packaging-sample
